numpy/core/numeric.py

```diff
@@ -1097,10 +1113,24 @@ def tensordot(a, b, axes=2):
     newshape_b = (N2, int(multiply.reduce([bs[ax] for ax in notin])))
     oldb = [bs[axis] for axis in notin]
 
+    if out is not None:
+        correct = True
+        if isinstance(out, np.ndarray):
+            if out.shape != tuple(olda+oldb):
+                correct = False
+            else:
+                out = out.reshape((newshape_a[0], newshape_b[1]))
+        else: 
+            correct = False
+        if correct is False: 
+            raise ValueError("output array is not acceptable (must have"
+            "the right datatype, number of dimensions, and be a C-Array)")
+
     at = a.transpose(newaxes_a).reshape(newshape_a)
     bt = b.transpose(newaxes_b).reshape(newshape_b)
-    res = dot(at, bt)
-    return res.reshape(olda + oldb)
+    
+    out = dot(at, bt, out)
+    return out.reshape(olda + oldb)
```

Comment:
This will end up returning a view of `out` when `out` is specified (since it is a reshaped version of the reshaped version). I am not sure I mind that though.

Reply:
I think we promise that `return(func, out=out) is out` for functions that have an `out` kwarg, no?

Reply:
@mattip: Indeed, although that promise is hard to keep unless `transpose` does not return a view. I think the logic would be:

* (somehow) check if transpose returns a view
* If it does, then return the original out array
* If it does not, then do a copy from the transposed array back to the original array.

Reply:
Can we rewrite this function to use `matmul` instead of `dot`, using the new `axes` argument to matmul? Then the `out` would be handled for us without any reshaping.

Reply:
One small issue with such rewriting is that it gives a different speed tradeoff currently. I would not actually mind the lazy way of doing `type(out) is np.ndarray` saying that subclasses may be allowed in the future.
The whole `__array_wrap__` dance is not fun, also the function currently uses `np.asarray` on the inputs in any case.